software/o_c_REV/HEM_VectorLFO.ino

```diff
@@ -104,26 +107,47 @@ public:
             ForEachChannel(ch) osc[ch].Reset();
         }
         if (c == 0) { // Frequency
-            if (freq[ch] > 100000) direction *= 10000;
-            else if (freq[ch] > 10000) direction *= 1000;
-            else if (freq[ch] > 1000) direction *= 100;
-            else if (freq[ch] > 300) direction *= 10;
-            freq[ch] = constrain(freq[ch] + direction, 10, 99900);
-            osc[ch].SetFrequency(freq[ch]);
+            int sign = (direction > 0) ? 1 : -1;
+
+            // if the encoder moved more than once, ensure we use the appropriate increment
+            //  when crossing boundaries
+            for (int i = 0; i < abs(direction); ++i) {
+              int cur_direction = sign;
+              if (freq[ch] + sign > 10000) cur_direction *= 1000;
+              else if (freq[ch] + sign > 1000) cur_direction *= 100;
+              else if (freq[ch] + sign > 250) cur_direction *= 10;  
+              freq[ch] = constrain(freq[ch] + cur_direction, min_freq, max_freq);
+            }
+            osc[ch].SetFrequency(freq[ch]);  
         }
     }
         
     uint32_t OnDataRequest() {
         uint32_t data = 0;
         Pack(data, PackLocation {0,6}, waveform_number[0]);
         Pack(data, PackLocation {6,6}, waveform_number[1]);
-        Pack(data, PackLocation {12,10}, freq[0] & 0x03ff);
-        Pack(data, PackLocation {22,10}, freq[1] & 0x03ff);
+
+        for (int i = 0; i < 2; ++i) {
+          int exponent = 0;
+          if (freq[i] > 250) exponent++;
+          if (freq[i] > 1000) exponent++;
+          if (freq[i] > 10000) exponent++;
+          Pack(data, PackLocation {12 + i * 10, 2}, exponent);
+
+          int mantissa = freq[i] / pow(10, exponent);
+          Pack(data, PackLocation {12 + i * 10 + 2, 8}, mantissa);
+        }
+        
         return data;
     }
+    
     void OnDataReceive(uint32_t data) {
-        freq[0] = Unpack(data, PackLocation {12,10});
-        freq[1] = Unpack(data, PackLocation {22,10});
+        for (int i = 0; i < 2; ++i) {
+          int exponent = Unpack(data, PackLocation {12 + i * 10, 2});
+          int mantissa = Unpack(data, PackLocation {12 + i * 10 + 2, 8});
+          
+          freq[i] = mantissa * pow(10, exponent);
```

Comment:
Teensy 3.1/3.2 doesn't have hardware support for floats or doubles, so you might want to avoid `pow`. Using `pow` is slow (nbd here since this is `OnDataReceive`) and, more importantly, increases program size a ton. Consider a simple lookup table, e.g.:

```
static constexpr int pow10_lut[] = { 1, 10, 100, 1000 };
```

and then

```
freq[i] = mantissa * pow10_lut[exponent];
```

Reply:
Oh! That's great! Thanks!